### PR TITLE
[Jenkins] Add delete environment pipeline

### PR DESCRIPTION
**Pull request by lewijacn**
### Description
Adds logic for deleting our CFN stacks across two CDK packages we currently use (this projects CDK and sample EC2 source CDK). This logic can be used locally with python, but is also included as a Jenkins pipeline in this change to allow manually deleting an environment as well as scheduled deletion of environments.

### Issues Resolved
https://opensearch.atlassian.net/browse/MIGRATIONS-2054

### Testing
Local testing and Jenkins: https://migrations.ci.opensearch.org/job/delete-environment/11/console

### Check List
- [ ] New functionality includes testing
  - [ ] All tests pass, including unit test, integration test and doctest
- [ ] New functionality has been documented
- [ ] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: test/cleanupDeployment/cleanup_deployment.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+from typing import List
+
+from botocore.exceptions import ClientError
+
+logging.basicConfig(format='%(asctime)s [%(levelname)s] %(message)s', level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+INDEPENDENT_STACKS = ['MigrationConsole', 'ReindexFromSnapshot', 'TrafficReplayer', 'TargetClusterProxy',
+                      'CaptureProxy', 'KafkaBroker', 'OpenSearchContainer', 'CaptureProxyES', 'Elasticsearch']
+CORE_STACKS_ORDERED = ['MigrationInfra', 'OpenSearchDomain', 'NetworkInfra', 'infra-stack', 'network-stack']
```
[the remark this answers is not in the document]
Comment (peternied) [reply test/cleanupDeployment/cleanup_deployment.py:15]: In these test accounts are there any stacks that shouldn't be fully deleted? If that is the case, can we just tear down everything

File: test/cleanupDeployment/cleanup_deployment.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+        next_list_stacks_response = client.list_stacks(NextToken=next_token)
+        next_stack_names = [stack['StackName'] for stack in next_list_stacks_response['StackSummaries']
+                            if stack['StackStatus'] not in CFN_INITIAL_STATUS_SKIP]
+        stack_names.extend(next_stack_names)
+        list_stacks_response.get("NextToken", None)
+
+    stage_stack_names = []
+    for name in stack_names:
+        if re.match(rf".*-{stage_name}-.*|.*-{stage_name}$", name):
+            stage_stack_names.append(name)
```
[the remark this answers is not in the document]
Comment (lewijacn) [reply test/cleanupDeployment/cleanup_deployment.py:133]: From our list of all stack names we use regex to get stacks that have stage name in middle(-stage-) or at end(-stage) of stack name. Added a comment for this as well

File: vars/cleanupDeployment.groovy [hunk 3]
```diff
@@ -0,0 +1,47 @@
+def call(Map config = [:]) {
+
+    pipeline {
```
[the remark this answers is not in the document]
Comment (lewijacn) [reply vars/cleanupDeployment.groovy:3]: A human operator can simply input a stage name and trigger the delete process through the UI. We can also in the future schedule a job to delete a specified set of stages

File: test/cleanupDeployment/cleanup_deployment.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+        logger.info(f"The following stacks have been deleted "
+                    f"successfully: {[s.stack_name for s in stack_delete_requests]}")
+
+
+def delete_stacks(cfn_client, stack_names):
+    # Delete independent stacks in batch
+    independent_stack_delete_requests = [
+        delete_stack(cfn_client, stack_name)
+        for stack_name in stack_names
+        if any(stack_id in stack_name for stack_id in INDEPENDENT_STACKS)
```
[the remark this answers is not in the document]
Comment (lewijacn) [reply test/cleanupDeployment/cleanup_deployment.py:100]: The` INDEPENDENT_STACKS` list contains they key identifier for a given stack while `stack_name` is the full stack name. The `if` condition is checking if any of the key identifiers are in the stack name before deleting it, as we only want to delete our independent stacks in this section

File: test/cleanupDeployment/cleanup_deployment.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+
+
+class StackDeletionRequest:
+    def __init__(self, stack_name, client_request_token=None):
+        self.stack_name = stack_name
+        self.client_request_token = client_request_token
+        self.retry_count = 0
+
+
+def delete_stack(cfn_client, stack_name: str) -> StackDeletionRequest:
```
[the remark this answers is not in the document]
Comment (lewijacn) [reply test/cleanupDeployment/cleanup_deployment.py:29]: I think we should be fine to continue if a stack has already been deleted before we go to delete it. I have added logging for this and raising an error if its for a different reason besides not existing

**Review by chelma — APPROVED** [review]

**Pull request merged**